### PR TITLE
Fix secret registry

### LINKED ISSUE #3552: Simplify and improve the Secret Registry proxy
## Tasks

1. Fix the unit tests
1. Handle race conditions for parallel secret registration
1. Make all greenlets wait the secret registration transaction of interest
1. Replace unrecoverable for recoverable errors

### PR DESCRIPTION
**Pull request by hackaugusto**
fix #3552 .

- [x] Fix the unit tests
- [x] Handle race conditions for parallel secret registration
- [x] Make all greenlets wait the secret registration transaction of interest
- [x] ~~Replace unrecoverable for recoverable errors~~ Any of the transactions failures in this proxy are unrecoverable, explained with comments in the code.

### CODE REVIEW

File: raiden/network/proxies/secret_registry.py [hunk 1]
```diff
@@ -88,61 +120,175 @@ def register_secret_batch(
...
+            )
+
+            log.debug('registerSecretBatch called', **log_details)
+
             try:
-                gas_limit = safe_gas_limit(
+                transaction_hash = self.proxy.transact(
+                    'registerSecretBatch',
                     gas_limit,
-                    len(secrets) * GAS_REQUIRED_PER_SECRET_IN_BATCH,
```
Comment (hackaugusto) [inline raiden/network/proxies/secret_registry.py]: This was a bug, we don't want to send `secrets` but `secrets_to_register`

File: raiden/network/proxies/secret_registry.py [hunk 2]
```diff
@@ -88,61 +120,175 @@ def register_secret_batch(
...
+                    secrets_to_register,
                 )
-                transaction_hash = self.proxy.transact('registerSecretBatch', gas_limit, secrets)
```
Comment (hackaugusto) [inline raiden/network/proxies/secret_registry.py]: This was a bug, it should have been `secrets_to_register`

File: raiden/network/proxies/secret_registry.py [hunk 3]
```diff
@@ -59,23 +70,44 @@ def register_secret_batch(
...
+                secrethash_hex = encode_hex(secrethash)
 
-            is_register_needed = (
-                not self.check_registered(secrethash, given_block_identifier) and
-                secret not in self.open_secret_transactions
-            )
-            if is_register_needed:
-                secrets_to_register.append(secret)
-                secrethashes_to_register.append(secrethash_hex)
-                self.open_secret_transactions[secret] = secret_registry_transaction
```
Comment (hackaugusto) [inline raiden/network/proxies/secret_registry.py]: This was a bug, without synchronization `check_registered` above context switches and different threads may try to register the same secret

File: raiden/network/proxies/secret_registry.py [hunk 4]
```diff
@@ -88,61 +120,175 @@ def register_secret_batch(
...
-                block = receipt_or_none['blockNumber']
-            else:
-                block = checking_block
+                transaction_result.set(transaction_hash)
+
+        # Clear `open_secret_transactions` regardless of the transaction being
+        # successfully executed or not.
+        with self._open_secret_transactions_lock:
+            for secret in secrets_to_register:
+                self.open_secret_transactions.pop(secret)
```
Comment (hackaugusto) [inline raiden/network/proxies/secret_registry.py]: Not doing the cleanup on failure was also a bug, because if without the flag `--unrecoverable-error-should-crash` the node would not crash and the `open_secret_transactions` would be left in a dirty state.

File: raiden/network/proxies/secret_registry.py [hunk 5]
```diff
@@ -88,61 +120,175 @@ def register_secret_batch(
...
+            # - The account is lacking funds to pay for the gas.
+            #
+            # Either of these is a bug. The contract does not use
+            # assert/revert, and the account should always be funded
+
+            # For the
             self.proxy.jsonrpc_client.check_for_insufficient_eth(
                 transaction_name='registerSecretBatch',
-                transaction_executed=transaction_executed,
-                required_gas=len(secrets) * GAS_REQUIRED_PER_SECRET_IN_BATCH,
```
Comment (hackaugusto) [inline raiden/network/proxies/secret_registry.py]: Again, same problem as before, now it just uses the same `gas_limit` that failed the transaction

File: raiden/network/proxies/secret_registry.py [hunk 6]
```diff
@@ -88,61 +120,175 @@ def register_secret_batch(
...
+                transaction_result.set_exception(e)
+                msg = f'Unexpected exception {e} at sending registerSecretBatch transaction.'
             else:
-                secret_registry_transaction.set(transaction_hash)
-            finally:
-                for secret in secrets_to_register:
-                    self.open_secret_transactions.pop(secret, None)
-
-        transaction_executed = gas_limit is not None
-        if not transaction_executed or receipt_or_none:
```
Comment (hackaugusto) [inline raiden/network/proxies/secret_registry.py]: This was a bug use before assignment bug,  `receipt_or_none` was not set if the transaction was not sent

File: raiden/network/proxies/secret_registry.py [hunk 7]
```diff
@@ -35,9 +40,11 @@ def __init__(
             to_normalized_address(secret_registry_address),
         )
 
+        # There should be only one smart contract deployed, to avoid race
+        # conditions for on-chain unlocks.
         compare_contract_versions(
             proxy=proxy,
-            expected_version=contract_manager.contracts_version,
+            expected_version='0.4.0',
```
Comment (LefterisJP) [inline raiden/network/proxies/secret_registry.py]: Hm why this change? We can use different smart contract versions. We also do so in tests. Some tests use the limit ones and some the normal ones. For example smoke tests fail in your PR and it's probably because they required the non-limit contracts as you can see [here](https://github.com/raiden-network/raiden/blob/a48d5d1f70093869fd235020cbe3d1cf4cebd20e/raiden/ui/cli.py#L560).

File: raiden/network/proxies/secret_registry.py [hunk 8]
```diff
@@ -59,23 +71,46 @@ def register_secret_batch(
...
-                secret not in self.open_secret_transactions
-            )
-            if is_register_needed:
-                secrets_to_register.append(secret)
-                secrethashes_to_register.append(secrethash_hex)
-                self.open_secret_transactions[secret] = secret_registry_transaction
-            else:
-                secrethashes_not_sent.append(secrethash_hex)
+                # Do the local test on `open_secret_transactions` first, then
+                # if necessary do a RPC call.
```
Comment (LefterisJP) [inline raiden/network/proxies/secret_registry.py]: nitpick: do an RPC call

File: raiden/network/proxies/secret_registry.py (hunk 1 above)
Comment (LefterisJP) [reply raiden/network/proxies/secret_registry.py]: arh nice catch for all the `secrets` -> `secrets_to_register` palces

File: raiden/network/proxies/secret_registry.py [hunk 9]
```diff
@@ -84,65 +119,187 @@ def register_secret_batch(
...
+            log.debug(
+                'registerSecretBatch skipped, waiting for transactions',
+                **log_details,
+            )
+            gevent.joinall(wait_for, raise_error=True)
             return
 
         checking_block = self.client.get_checking_block()
-        error_prefix = 'Call to registerSecretBatch will fail'
         gas_limit = self.proxy.estimate_gas(checking_block, 'registerSecretBatch', secrets)
```
Comment (LefterisJP) [inline raiden/network/proxies/secret_registry.py]: Shouldn't you also change `secrets` to `secrets_to_register` here too?

File: raiden/network/proxies/secret_registry.py [hunk 10]
```diff
@@ -84,65 +119,187 @@ def register_secret_batch(
...
+            # gas_limit can fail because:
+            #
+            # - The Ethereum client detected the transaction could not
+            # successfully execute, this happens if an assert/revert is hit.
+            # - The account is lacking funds to pay for the gas.
+            #
+            # Either of these is a bug. The contract does not use
+            # assert/revert, and the account should always be funded
+
+            # For the
```
Comment (LefterisJP) [inline raiden/network/proxies/secret_registry.py]: For the ..? I think you did not finish this comment.

File: raiden/network/proxies/secret_registry.py [hunk 11]
```diff
@@ -84,65 +119,187 @@ def register_secret_batch(
...
+        registered, None if the secret was never registered.
+        """
+        result = self.proxy.contract.functions.getSecretRevealBlockHeight(
             secrethash,
         ).call(block_identifier=block_identifier)
 
-    def check_registered(
+        # Block 0 either represents the genesis block or an empty entry in the
+        # secret mapping. This is important for custom genesis files used while
+        # testing, to avoid problems the smart contract can be added as part of
```
Comment (LefterisJP) [inline raiden/network/proxies/secret_registry.py]: nitpick: testing. To ..

File: raiden/network/proxies/secret_registry.py [hunk 12]
```diff
@@ -84,65 +119,187 @@ def register_secret_batch(
...
+        # the genesis file, however it's important for it's storage to be
```
Comment (LefterisJP) [inline raiden/network/proxies/secret_registry.py]: typo: it's storage -> its storage

File: raiden/tests/integration/network/proxies/test_secret_registry.py [hunk 13]
```diff
@@ -0,0 +1,212 @@
+from collections import defaultdict
```
Comment (LefterisJP) [inline raiden/tests/integration/network/proxies/test_secret_registry.py]: What's the logic of moving the contract proxy tests from `integration/contracts/ to `integration/network/proxies/` ? To match the directory structure of the client?

File: raiden/network/proxies/secret_registry.py [hunk 14]
```diff
@@ -84,65 +119,187 @@ def register_secret_batch(
...
+            # - The nonce of the transaction was already used
+            # - The nonce was reused *and* the account didn't have enough ether
+            # to pay for the gas
+            #
+            # Safety cannot be guaranteed under any of these cases, this error
+            # is unrecoverable. *Note*: This assumes the ethereum client
+            # takes into account the current transactions in the pool.
+            if gas_limit:
+                assert msg, (
+                    'Unexpected control flow, an exception should habe been raised.'
```
Comment (rakanalh) [inline raiden/network/proxies/secret_registry.py]: s/habe/have

File: raiden/tests/integration/network/proxies/test_secret_registry.py [hunk 15]
```diff
@@ -0,0 +1,212 @@
...
+
+        block = secret_registry_proxy.get_secret_registration_block_by_secrethash(
+            secrethash=secrethash,
+            block_identifier='latest',
+        )
+        msg = 'Block number reported by the proxy and the event must match'
+        assert block == secret_registered['args']['blockNumber'], msg
+
+
+def test_concurrent_secret_registartion(secret_registry_proxy, monkeypatch):
```
Comment (rakanalh) [inline raiden/tests/integration/network/proxies/test_secret_registry.py]: s/registartion/registration

File: raiden/network/proxies/secret_registry.py (hunk 7 above)
Comment (hackaugusto) [reply raiden/network/proxies/secret_registry.py]: The rationale is: We should only have one secret registry in the mainnet, if we do have more than one registry things will get unnecessarily complicate because of race conditions. The tests should just use the version of the smart contract that is deployed on the mainnet, otherwise we may be testing against a different behavior.

The problem is that new packages of the raiden-contracts are bumping the version of the smart contracts even if there was no change, so in theory the only different of `0.4.0` to the `0.8.0` is the version number.

I will revert this change, but IMO we should never deploy a new secret registry on the mainnet, maybe the check can be done on a per chain id bases together by the contract address.

File: raiden/tests/integration/network/proxies/test_secret_registry.py (hunk 13 above)
Comment (hackaugusto) [reply raiden/tests/integration/network/proxies/test_secret_registry.py]: Yes, otherwise it's unnecessarily hard to find the matching tests

**Comment by luehrsFred** [comment]
I experienced an error here on secret_registry.py in line [113](https://github.com/raiden-network/raiden/blob/master/raiden/network/proxies/secret_registry.py#L113)

`UnboundLocalError: local variable 'receipt_or_none' referenced before assignment`

[raiden-exception-2019-03-05T18-44zv3r5owb.txt](https://github.com/raiden-network/raiden/files/2935303/raiden-exception-2019-03-05T18-44zv3r5owb.txt)

My Node was trying to register secrets but was not funded with ETH. If you prefer I can open a complete issue to that. lmk

**Comment by hackaugusto** [comment]
> My Node was trying to register secrets but was not funded with ETH. If you prefer I can open a complete issue to that. lmk

This error is fixed by the PR.
